### ISSUE TITLE
Headless api fails to launch on server.

### ISSUE THREAD

**Issue by citkane** [mrosseel/bisq-api#22]
All is working on a laptop/desktop when running the headless API, but on an actual headless server:
```
BISQ_API_PORT=4445 mvn exec:java -X -Dexec.mainClass="io.bisq.api.app.ApiMain" -Dexec.args="--baseCurrencyNetwork=BTC_REGTEST --bitcoinRegtestHost localhost --nodePort 4444 --useLocalhost true --appName Bisq-Regtest-Arbitrator --seedNodes=localhost:2002"
```
Produces: [server.log](https://github.com/mrosseel/bisq-api/files/1704566/server.log)

What appears to be the relevant bit is:
```
java.lang.UnsupportedOperationException: Unable to open DISPLAY
    at com.sun.glass.ui.gtk.GtkApplication.<init> (GtkApplication.java:68)
    at com.sun.glass.ui.gtk.GtkPlatformFactory.createApplication (GtkPlatformFactory.java:41)
    at com.sun.glass.ui.Application.run (Application.java:146)
    at com.sun.javafx.tk.quantum.QuantumToolkit.startup (QuantumToolkit.java:257)
    at com.sun.javafx.application.PlatformImpl.startup (PlatformImpl.java:211)
    at javafx.embed.swing.JFXPanel.initFx (JFXPanel.java:215)
    at javafx.embed.swing.JFXPanel.<init> (JFXPanel.java:230)
    at io.bisq.api.app.MainViewModelHeadless.lambda$initialiseJavaFxToolkit$0 (MainViewModelHeadless.java:274)
    at java.awt.event.InvocationEvent.dispatch (InvocationEvent.java:311)
    at java.awt.EventQueue.dispatchEventImpl (EventQueue.java:756)
    at java.awt.EventQueue.access$500 (EventQueue.java:97)
    at java.awt.EventQueue$3.run (EventQueue.java:709)
    at java.awt.EventQueue$3.run (EventQueue.java:703)
    at java.security.AccessController.doPrivileged (Native Method)
    at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege (ProtectionDomain.java:80)
    at java.awt.EventQueue.dispatchEvent (EventQueue.java:726)
    at java.awt.EventDispatchThread.pumpOneEventForFilters (EventDispatchThread.java:201)
    at java.awt.EventDispatchThread.pumpEventsForFilter (EventDispatchThread.java:116)
    at java.awt.EventDispatchThread.pumpEventsForHierarchy (EventDispatchThread.java:105)
    at java.awt.EventDispatchThread.pumpEvents (EventDispatchThread.java:101)
    at java.awt.EventDispatchThread.pumpEvents (EventDispatchThread.java:93)
    at java.awt.EventDispatchThread.run (EventDispatchThread.java:82)
Feb-07 21:04:08.672 [BisqApiMain] ERROR i.b.api.app.Api: Uncaught Exception from thread BisqApiMain 
Feb-07 21:04:08.673 [BisqApiMain] ERROR i.b.api.app.Api: throwableMessage= null 
Feb-07 21:04:08.673 [BisqApiMain] ERROR i.b.api.app.Api: throwableClass= class java.lang.ExceptionInInitializerError 
Feb-07 21:04:08.676 [BisqApiMain] ERROR i.b.api.app.Api: Stack trace:
java.lang.ExceptionInInitializerError
	at io.bisq.api.app.MainViewModelHeadless.initialiseJavaFxToolkit(MainViewModelHeadless.java:279)
	at io.bisq.api.app.MainViewModelHeadless.start(MainViewModelHeadless.java:253)
	at io.bisq.api.app.Api.<init>(Api.java:119)
	at io.bisq.api.app.ApiMain.lambda$doExecute$0(ApiMain.java:95)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
 
java.lang.ExceptionInInitializerError
	at io.bisq.api.app.MainViewModelHeadless.initialiseJavaFxToolkit(MainViewModelHeadless.java:279)
	at io.bisq.api.app.MainViewModelHeadless.start(MainViewModelHeadless.java:253)
	at io.bisq.api.app.Api.<init>(Api.java:119)
	at io.bisq.api.app.ApiMain.lambda$doExecute$0(ApiMain.java:95)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by mrosseel** [comment]
yes that might be a problem, I have to actually start the javafx framework. To avoid this, heavy refactoring in bisq will be needed I think. 

For the time being, it might be a solution to install X on your remote server. Let me know if this solves it

**Comment by citkane** [comment]
Yes, heavy refactoring is required to achieve a scalable, dev friendly MVC pattern. At the moment the GUI logic is entangled with the business logic.

I have started the work here: https://github.com/citkane/bisq-engine
At the moment this is just plumbing from app to API, but it runs from jar, connects to the network and serves responses in true headless mode.

I am hesitant to install X on my server - it feels like a hackish workaround that can only lead to tears at the bottom of a rabbit-hole :)

**Comment by mrosseel** [comment]
@citkane not sure what your plan is with bisq-engine, the best approach is to refactor bisq-gui so that, as you say, business logic and GUI are strictly separated. At that point, bisq-api can run purely headless. Starting from scratch like you did in bisq-engine while take a lot of time to get it right.

Regarding X: another possibility is to run it inside of docker, like that you can install X inside of the container. Someone is writing those docker containers for integration testing so should be done pretty soon.

Proposal: do the dirty deed and use X on the server (either native or in docker) and we'll discuss how best to remove that dependency. Open for hangout/... to discuss.

**Comment by citkane** [comment]
> the best approach is to refactor bisq-gui so that, as you say, business logic and GUI are strictly separated

This is my plan with bisq-engine. It is the Model and Controller of MVC, wrapped in Spring-boot as a convenience to integrate http. It will provide (agnostic) CRUD endpoints, watchers and event broadcasting. 

This leaves bisq-gui as the View of MVC, and the refactoring happens here to interface with an agnostic CRUD. Whether these are one or two modules of Bisq-Network is semantic for now. I have split them for my own sanity and good order, but the goal is that they work together and can be packaged as one.

My personal approach is going to be to dissect GUI is it stands to understand what it needs and how it thinks. This will inform the best approach for a design pattern for CRUD, which theoretically should integrate the work already done on bisq-api and existing GUI data formatters.

The goal here is re-usability and common data formatting for internal, external VIEWs and 3rd party API's, ultimately facilitating a broader BISQ design pattern as proposed here: https://github.com/bisq-network/proposals/issues/3#issuecomment-361959884

I am of the feeling that dirty deeds, hacks and workarounds now - while the API has no production dependencies - indicates that the underlying architecture needs to be addressed. From what I am seeing, this needs to include two major work components:

1. Design and implementation of data endpoints 
2. Refactoring the javaFX gui application to interact with these endpoints 

I want to do some more work, and understand existing approaches better, before being too emphatic. For now however I am going to park https://github.com/citkane/bisq-front until a best approach emerges.

I am on GMT, but normally work into the evenings. I am happy to `hangout/...` if you propose a time and the best platform for communication. I assume you you are meaning the Google hangouts?

**Comment by mrosseel** [comment]
@citkane yes I was referring to google hangouts, but have been using jitsi lately. Also on GMT, how about we try to talk tomorrow, Wednesday 21 Feb? I'm available between 8:00 till 16:00 and from 18:00 till 22:00, you choose :)

**Comment by citkane** [comment]
@mrosseel I will try and be available from 18:00 tomorrow 21 Feb, but there is a small chance I won't. I am dealing with a terminally sick pet and upset family :(

We can link with https://meet.jit.si/mroseelcitkane

I have been making progress with [bisq-engine](https://github.com/citkane/bisq-engine)

It can start in headless or GUI mode, has an integrated Swagger API with UI and it is possible (through the API) to start with clean boot, accept TAC, get wallet funding address, fund wallet, create a SEPA account and create offers.

Give it a spin. All the API classes are in https://github.com/citkane/bisq-engine/tree/master/engine/src/main/java/io/bisq/engine/app/api 

The design pattern should make it a lot easier to focus on API development without getting tangled in UI stuff.

**Comment by mrosseel** [comment]
@citkane no worries, let me know if you can make it if not we'll find another time

Can you tell me what's still missing on bisq-engine?

There's a few PR's I have to review so won't be able to look at it today. The content of those PR's might interest you as well, bernard made an integration testing framework using docker containers, should help in detecting regressions wrt bisq.

**Comment by citkane** [comment]
Ok, how about this Friday at 2pm?

Best you have also had time to look at bisq-engine. It is not a matter of what is still missing, but a matter of the architecture that is there to build from:

- It separates the API from UI issues, so headless/GUI boot is optional and functional
- It uses Spring Swagger integration library Springfox, so the whole construct is within the Spring workflow with annotation integration.
- API has full access to all the low level bisq commands from the injector (not via UI), so it is possible to build out any logic and functionality.
- Pre-existing business logic from UI module dataModels can be 'borrowed' and re-used through interfaces (this is still messy, but can easily be cleaned up without much re-factoring in the UI module)

I am working through a basic workflow for the API, and here is where I am:
- Boot from scratch in headless with wallet initialised (done)
- Accept TAC (done)
- Get wallet funding address and fund wallet (done)
- Create an account (done SEPA)
- Create an offer (done)
- Accept an offer (busy with)
- Manage a trade (todo next)

**Comment by mrosseel** [comment]
@citkane yes friday 2pm should work
I'll have a look at your work, sounds like you did already quite a few things.
In my viewt the core of the UI issue is the coupling between UI and core functionality in bisq-exchange. So the only way to solve this cleanly is to refactor bisq-exchange so it exposes all functionality needed by the API. 
At the moment in the bisq-api there is a 'fat' BisqProxy class which does all kinds of stuff to make it work. Once bisq-exchange refactoring is done, this proxy layer can become very thin.
In short - if your proposal is to refactor bisq-exchange, I'm all for it. If you want to copy code from bisq-exchange to  bisq-engine and make it cleaner, I fear it will be a never-ending game of catchup without fixing the real problem. 
My goal with bisq-api was to present a workable api, so people would build stuff on top of it. The internals might be a bit ugly at the moment, but an API user shouldn't care. Yes there might be a need to install X on your VPS, but that's the only leaking of ugliness at the moment. 

PS: once there is a useful app built on top of the API's, there's also more incentive to refactor bisq-exchange. As it is now, there are 0 working API apps, so refactoring bisq-exchange and introducing new bugs etc etc might not seem worth it to the other contributors. That's why I'm more excited for your frontend project in the short term, but will certainly look at your project and see what I can learn from it!

**Comment by mrosseel** [comment]
@citkane just to repeat, I'll be in the meeting at 2pm GMT; we now have a 'standard' meeting room to discuss api stuff here: [https://meet.jit.si/BisqTradingApiMeeting](https://meet.jit.si/BisqTradingApiMeeting)
 Probably bernard will also join, he did the integration testing framework, so this will be the first time everyone working on bisq-api will have a meeting! Let's discuss and see how we can move forward, I'll read your code before the code so I know better what I'm talking about ;)

**Comment by citkane** [comment]
Great, speak to you at 2.

I pushed to engine a minute ago the ability to accept an offer.

I appreciate your goal, but also see the frustration and effort in working with the ugliness. The exchange stack really is independent of the UI and, once booted into the correct context, works independently of the UI.

The UI module `datamodel` classes then present the methods that assemble the business logic for the API and UI both without having to launch the JavaFX application (with very minimal refactoring needed to prevent popups in headless).

This all works. I will show you later today.

**Comment by blabno** [comment]
Since 36537965fe91bb8302d155b1963e513db7350170 the api works inside docker without xvfb, so it's truly headless.

**Comment by Nikhil-cyber-93** [comment]
@mrosseel,

Hi sir. I am getting the following error when I try to run QuPath application (which is in remote pc) from my home system. However, QuPath in my pc works perfectly fine . What could be the issue??

Exception in thread "main" java.lang.UnsupportedOperationException: Unable to open DISPLAY
	at com.sun.glass.ui.gtk.GtkApplication.lambda$new$6(GtkApplication.java:173)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at com.sun.glass.ui.gtk.GtkApplication.<init>(GtkApplication.java:171)
	at com.sun.glass.ui.gtk.GtkPlatformFactory.createApplication(GtkPlatformFactory.java:41)
	at com.sun.glass.ui.Application.run(Application.java:144)
	at com.sun.javafx.tk.quantum.QuantumToolkit.startup(QuantumToolkit.java:280)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:269)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:158)
	at com.sun.javafx.application.LauncherImpl.startToolkit(LauncherImpl.java:658)
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:678)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
	at java.base/java.lang.Thread.run(Unknown Source)
QuPath-0.2.3 Failed to launch JVM
